lib/synapse/config_generator/base.rb

```diff
@@ -0,0 +1,37 @@
+require 'synapse/log'
+
+class Synapse::ConfigGenerator
+  class BaseGenerator
+    include Synapse::Logging
+
+    NAME = 'base'.freeze
+
+    # Exposes NAME as 'name' so we can remain consistent with how we refer to
+    # service_watchers' by generator.name access (instead of
+    # generator.class::NAME) even though the names of generators don't change
+    def name
+      self.class::NAME
+    end
+
+    # The synapse main loop will call this any time watchers change, the
+    # config_generator is responsible for diffing the passed watcher state
+    # against the output configuration
+    def update_config(watchers)
+    end
+
+    # The synapse main loop will call this every tick
+    # of the logical clock (~1s). You can use this to intiate reloads
+    # or restarts in a rate limited fashion
+    def tick
+    end
+
+    # Service watchers have a subsection of their ``services`` entry that is
+    # dedicated to the watcher specific configuration for how to configure
+    # the config generator. This method will be called with each of these
+    # watcher hashes, and should normalize them to what the config generator
+    # needs, such as adding defaults. Mutate the passed hash
+    def normalize_config_generator_opts!(service_watcher_name, service_watcher_opts)
```

Comment:
> i was confused by there being two sources of options for a config generator. the first is the options passed to the initialize method, which contain overall options. the second is the watcher-specific options for this config generator.

Yup, the fact that there is coupling between the two presented a significant challenge during this refactor.

>maybe we should call these options watcher_opts rather than config_generator_opts to clarify the distinction?
Hm, I will work on making this clearer

> also, should we really be encouraging side-effects? this might lead to trouble. given how we actually use this code below, returning a copy of the passed-in hash would work just as well. we should encourage users to do something like:

Yea, I think I can figure that out

>finally, i was also confused by the fact that we have to reach into the watchers to get the config generator options each time we generate a config for that watcher. i guess i'm not sure what to do about that problem, but i'm kind of annoyed about repeated calls to, e.g., watcher.generator_opts['haproxy']

Yes, I agree, but it's not that bad and we can refactor this in the future imo.